File: bindings/python/src/datalevin/_raw.py
"""Advanced raw access to the Datalevin interop bridge."""

from __future__ import annotations

from ._interop import _BINDINGS


class RawInterop:
    """Expose raw `DatalevinInterop` operations for advanced callers."""

    def api_info(self):
        return _BINDINGS.api_info_raw()

    def exec_json(self, request_json: str):
        return _BINDINGS.exec_json_raw(request_json)

    def core_invoke(self, function: str, args=None):
        return _BINDINGS.core_invoke(function, args)

    def client_invoke(self, function: str, args=None):
        return _BINDINGS.client_invoke(function, args)

    def create_connection(self, dir=None, schema=None, opts=None, *, shared: bool = False):
        return _BINDINGS.create_connection(dir, schema, opts, shared=shared)

    def init_db(self, datoms, dir=None, schema=None, opts=None):
        return _BINDINGS.init_db(datoms, dir, schema, opts)

    def fill_db(self, conn, datoms):
        return _BINDINGS.fill_db(conn, datoms)

    def close_connection(self, handle):
        _BINDINGS.close_connection(handle)

    def connection_closed(self, handle):
        return _BINDINGS.connection_closed(handle)

    def connection_db(self, handle):
        return _BINDINGS.connection_db(handle)

    def connection_datalog_kv(self, handle):
        return _BINDINGS.connection_datalog_kv(handle)

    def connection_datoms(self, handle, index, c1=None, c2=None, c3=None, limit=None):
        return _BINDINGS.connection_datoms(handle, index, c1, c2, c3, limit)

    def connection_search_datoms(self, handle, e=None, attr=None, value=None):
        return _BINDINGS.connection_search_datoms(handle, e, attr, value)

    def connection_count_datoms(self, handle, e=None, attr=None, value=None):
        return _BINDINGS.connection_count_datoms(handle, e, attr, value)

    def connection_seek_datoms(self, handle, index, c1=None, c2=None, c3=None, limit=None):
        return _BINDINGS.connection_seek_datoms(handle, index, c1, c2, c3, limit)

    def connection_rseek_datoms(self, handle, index, c1=None, c2=None, c3=None, limit=None):
        return _BINDINGS.connection_rseek_datoms(handle, index, c1, c2, c3, limit)

    def connection_index_range(self, handle, attr, start, end):
        return _BINDINGS.connection_index_range(handle, attr, start, end)

    def connection_fulltext_datoms(self, handle, query, opts=None):
        return _BINDINGS.connection_fulltext_datoms(handle, query, opts)

    def connection_copy(self, handle, dest, compact=None):
        return _BINDINGS.connection_copy(handle, dest, compact)

    def connection_tx_log_watermarks(self, handle):
        return _BINDINGS.connection_tx_log_watermarks(handle)

    def connection_open_tx_log(self, handle, from_lsn, upto_lsn=None):
        return _BINDINGS.connection_open_tx_log(handle, from_lsn, upto_lsn)

    def connection_create_snapshot(self, handle):
        return _BINDINGS.connection_create_snapshot(handle)

    def connection_list_snapshots(self, handle):
        return _BINDINGS.connection_list_snapshots(handle)

    def connection_gc_tx_log_segments(self, handle, retain_floor_lsn=None):
        return _BINDINGS.connection_gc_tx_log_segments(handle, retain_floor_lsn)

    def open_key_value(self, dir, opts=None):
        return _BINDINGS.open_key_value(dir, opts)

    def close_key_value(self, handle):
        _BINDINGS.close_key_value(handle)

    def key_value_closed(self, handle):
        return _BINDINGS.key_value_closed(handle)

    def new_client(self, uri, opts=None):
        return _BINDINGS.new_client(uri, opts)

    def close_client(self, handle):
        _BINDINGS.close_client(handle)

    def client_disconnected(self, handle):
        return _BINDINGS.client_disconnected(handle)

    def read_edn(self, edn: str):
        return _BINDINGS.read_edn(edn)

    def keyword(self, value: str):
        return _BINDINGS.keyword(value)

    def symbol(self, value: str):
        return _BINDINGS.symbol(value)

    def schema(self, schema):
        return _BINDINGS.schema(schema)

    def options(self, opts):
        return _BINDINGS.options(opts)

    def udf_descriptor(self, descriptor):
        return _BINDINGS.udf_descriptor(descriptor)

    def create_udf_registry(self):
        return _BINDINGS.create_udf_registry()

    def register_udf(self, registry, descriptor, fn):
        return _BINDINGS.register_udf(registry, descriptor, fn)

    def unregister_udf(self, registry, descriptor):
        return _BINDINGS.unregister_udf(registry, descriptor)

    def registered_udf(self, registry, descriptor):
        return _BINDINGS.registered_udf(registry, descriptor)

    def rename_map(self, rename_map):
        return _BINDINGS.rename_map(rename_map)

    def delete_attrs(self, attrs):
        return _BINDINGS.delete_attrs(attrs)

    def lookup_ref(self, value):
        return _BINDINGS.lookup_ref(value)

    def datom(self, e, attr, value, tx=None, added=None):
        return _BINDINGS.datom(e, attr, value, tx, added)

    def tx_data(self, tx_data):
        return _BINDINGS.tx_data(tx_data)

    def kv_txs(self, txs):
        return _BINDINGS.kv_txs(txs)

    def kv_type(self, value):
        return _BINDINGS.kv_type(value)

    def database_type(self, value: str):
        return _BINDINGS.database_type(value)

    def role(self, role: str):
        return _BINDINGS.role(role)

    def permission_keyword(self, value: str):
        return _BINDINGS.permission_keyword(value)

    def permission_target(self, object_type: str, target):
        return _BINDINGS.permission_target(object_type, target)


_RAW_INTEROP = RawInterop()


def interop() -> RawInterop:
    """Return the singleton raw interop facade."""

    return _RAW_INTEROP

File: bindings/javascript/src/raw.js
import { CONNECTION_WITH_TRANSACTION_UNSUPPORTED, _BINDINGS } from "./interop.js";
import { toJava } from "./convert.js";
import { callJavaMethod } from "./jvm.js";
import { toJsResult } from "./result.js";

function datomTuple(e, attr, value, tx = undefined, added = undefined) {
  if (added !== undefined && tx === undefined) {
    throw new TypeError("tx is required when added is provided");
  }
  if (added !== undefined) {
    return [e, attr, value, tx, added];
  }
  if (tx !== undefined) {
    return [e, attr, value, tx];
  }
  return [e, attr, value];
}

function resourceHandle(value) {
  if (typeof value?.rawHandle === "function") {
    return value.rawHandle();
  }
  return value;
}

function hasValue(value) {
  return value !== null && value !== undefined;
}

class RawInterop {
  async apiInfo() {
    return _BINDINGS.apiInfoRaw();
  }

  async execJson(op, args = null) {
    return _BINDINGS.execJson(op, args);
  }

  async coreInvoke(functionName, args = null) {
    return _BINDINGS.coreInvoke(functionName, args);
  }

  async clientInvoke(functionName, args = null) {
    return _BINDINGS.clientInvoke(functionName, args);
  }

  async createConnection(dir = null, schema = null, opts = null, { shared = false } = {}) {
    return _BINDINGS.createConnection(dir, schema, opts, { shared });
  }

  async initDb(datoms, dir = null, schema = null, opts = null) {
    return _BINDINGS.initDb(datoms, dir, schema, opts);
  }

  async fillDb(conn, datoms) {
    return _BINDINGS.fillDb(conn, datoms);
  }

  async closeConnection(handle) {
    await _BINDINGS.closeConnection(resourceHandle(handle));
  }

  async connectionClosed(handle) {
    return _BINDINGS.connectionClosed(resourceHandle(handle));
  }

  async connectionDb(handle) {
    const { Database } = await import("./database.js");
    return new Database(await _BINDINGS.connectionDb(resourceHandle(handle)));
  }

  async connectionEntity(handle, eid) {
    return _BINDINGS.connectionEntity(resourceHandle(handle), eid);
  }

  async databaseEntid(db, eid) {
    return _BINDINGS.databaseEntid(resourceHandle(db), eid);
  }

  async databaseEntity(db, eid) {
    return _BINDINGS.databaseEntity(resourceHandle(db), eid);
  }

  async databaseEntityMap(db, eid) {
    return toJsResult(await _BINDINGS.databaseEntityMap(resourceHandle(db), eid), { bridge: true });
  }

  async databasePull(db, selector, eid) {
    return toJsResult(await _BINDINGS.databasePull(resourceHandle(db), selector, eid), { bridge: true });
  }

  async databasePullMany(db, selector, eids) {
    return toJsResult(await _BINDINGS.databasePullMany(resourceHandle(db), selector, eids), { bridge: true });
  }

  async connectionTxDataToSimulatedReport(handle, txData) {
    return _BINDINGS.connectionTxDataToSimulatedReport(resourceHandle(handle), txData);
  }

  async entityIs(value) {
    return _BINDINGS.entityIs(value);
  }

  async entityId(entity) {
    return _BINDINGS.entityId(resourceHandle(entity));
  }

  async entityGet(entity, attr) {
    return _BINDINGS.entityGet(resourceHandle(entity), attr);
  }

  async entityContains(entity, attr) {
    return _BINDINGS.entityContains(resourceHandle(entity), attr);
  }

  async entityTouch(entity) {
    return toJsResult(await _BINDINGS.entityTouch(resourceHandle(entity)), { bridge: true });
  }

  async connectionDatalogKv(handle) {
    return callJavaMethod(resourceHandle(handle), "datalogKV");
  }

  async connectionTransactAsync(handle, txData, txMeta = null) {
    const future = await _BINDINGS.connectionTransactAsync(resourceHandle(handle), txData, txMeta);
    return toJsResult(await callJavaMethod(future, "get"), { bridge: true });
  }

  async connectionDatoms(handle, index, c1 = null, c2 = null, c3 = null, limit = null) {
    const args = [resourceHandle(handle), "datoms", await toJava(index), await toJava(c1), await toJava(c2), await toJava(c3)];
    if (hasValue(limit)) {
      args.push(await toJava(limit));
    }
    return toJsResult(await callJavaMethod(...args), { bridge: true });
  }

  async connectionSearchDatoms(handle, e = null, attr = null, value = null) {
    return toJsResult(
      await callJavaMethod(
        resourceHandle(handle),
        "searchDatoms",
        await toJava(e),
        await toJava(attr),
        await toJava(value)
      ),
      { bridge: true }
    );
  }

  async connectionCountDatoms(handle, e = null, attr = null, value = null) {
    return toJsResult(
      await callJavaMethod(
        resourceHandle(handle),
        "countDatoms",
        await toJava(e),
        await toJava(attr),
        await toJava(value)
      )
    );
  }

  async connectionSeekDatoms(handle, index, c1 = null, c2 = null, c3 = null, limit = null) {
    const args = [
      resourceHandle(handle),
      "seekDatoms",
      await toJava(index),
      await toJava(c1),
      await toJava(c2),
      await toJava(c3)
    ];
    if (hasValue(limit)) {
      args.push(await toJava(limit));
    }
    return toJsResult(await callJavaMethod(...args), { bridge: true });
  }

  async connectionRseekDatoms(handle, index, c1 = null, c2 = null, c3 = null, limit = null) {
    const args = [
      resourceHandle(handle),
      "rseekDatoms",
      await toJava(index),
      await toJava(c1),
      await toJava(c2),
      await toJava(c3)
    ];
    if (hasValue(limit)) {
      args.push(await toJava(limit));
    }
    return toJsResult(await callJavaMethod(...args), { bridge: true });
  }

  async connectionIndexRange(handle, attr, start, end) {
    return toJsResult(
      await callJavaMethod(
        resourceHandle(handle),
        "indexRange",
        await toJava(attr),
        await toJava(start),
        await toJava(end)
      ),
      { bridge: true }
    );
  }

  async connectionFulltextDatoms(handle, query, opts = null) {
    const result = hasValue(opts)
      ? await callJavaMethod(resourceHandle(handle), "fulltextDatoms", query, await toJava(opts))
      : await callJavaMethod(resourceHandle(handle), "fulltextDatoms", query);
    return toJsResult(result, { bridge: true });
  }

  async connectionCopy(handle, dest, compact = null) {
    if (hasValue(compact)) {
      return callJavaMethod(resourceHandle(handle), "copy", dest, Boolean(compact));
    }
    return callJavaMethod(resourceHandle(handle), "copy", dest);
  }

  async connectionTxLogWatermarks(handle) {
    return toJsResult(await callJavaMethod(resourceHandle(handle), "txLogWatermarks"), { bridge: true });
  }

  async connectionOpenTxLog(handle, fromLsn, uptoLsn = null) {
    const result = hasValue(uptoLsn)
      ? await callJavaMethod(resourceHandle(handle), "openTxLog", await toJava(fromLsn), await toJava(uptoLsn))
      : await callJavaMethod(resourceHandle(handle), "openTxLog", await toJava(fromLsn));
    return toJsResult(result, { bridge: true });
  }

  async connectionCreateSnapshot(handle) {
    return toJsResult(await callJavaMethod(resourceHandle(handle), "createSnapshot"), { bridge: true });
  }

  async connectionListSnapshots(handle) {
    return toJsResult(await callJavaMethod(resourceHandle(handle), "listSnapshots"), { bridge: true });
  }

  async connectionGcTxLogSegments(handle, retainFloorLsn = null) {
    if (hasValue(retainFloorLsn)) {
      return toJsResult(
        await callJavaMethod(resourceHandle(handle), "gcTxLogSegments", await toJava(retainFloorLsn)),
        { bridge: true }
      );
    }
    return toJsResult(await callJavaMethod(resourceHandle(handle), "gcTxLogSegments"), { bridge: true });
  }

  async connectionWithTransaction(handle, fn) {
    throw new DatalevinError(CONNECTION_WITH_TRANSACTION_UNSUPPORTED);
  }

  async connectionReIndex(handle, schema = null, opts = null) {
    return _BINDINGS.connectionReIndex(resourceHandle(handle), schema, opts);
  }

  async openKeyValue(dir, opts = null) {
    return _BINDINGS.openKeyValue(dir, opts);
  }

  async closeKeyValue(handle) {
    await _BINDINGS.closeKeyValue(resourceHandle(handle));
  }

  async keyValueClosed(handle) {
    return _BINDINGS.keyValueClosed(resourceHandle(handle));
  }

  async keyValueBeginTransaction(handle) {
    return _BINDINGS.keyValueBeginTransaction(resourceHandle(handle));
  }

  async keyValueCommitTransaction(tx) {
    return toJsResult(await _BINDINGS.keyValueCommitTransaction(resourceHandle(tx)));
  }

  async keyValueAbortTransaction(tx) {
    return toJsResult(await _BINDINGS.keyValueAbortTransaction(resourceHandle(tx)));
  }

  async keyValueWithTransaction(handle, fn) {
    return toJsResult(await _BINDINGS.keyValueWithTransaction(resourceHandle(handle), fn), { bridge: true });
  }

  async keyValueReIndex(handle, opts = null) {
    return _BINDINGS.keyValueReIndex(resourceHandle(handle), opts);
  }

  async newSearchEngine(kv, opts = null) {
    return _BINDINGS.newSearchEngine(resourceHandle(kv), opts);
  }

  async searchAddDoc(search, docRef, docText, checkExist = null) {
    return _BINDINGS.searchAddDoc(resourceHandle(search), docRef, docText, checkExist);
  }

  async searchRemoveDoc(search, docRef) {
    return _BINDINGS.searchRemoveDoc(resourceHandle(search), docRef);
  }

  async searchClearDocs(search) {
    return _BINDINGS.searchClearDocs(resourceHandle(search));
  }

  async searchDocIndexed(search, docRef) {
    return _BINDINGS.searchDocIndexed(resourceHandle(search), docRef);
  }

  async searchDocCount(search) {
    return toJsResult(await _BINDINGS.searchDocCount(resourceHandle(search)));
  }

  async search(search, query, opts = null) {
    return toJsResult(await _BINDINGS.search(resourceHandle(search), query, opts));
  }

  async searchReIndex(search, opts = null) {
    return _BINDINGS.searchReIndex(resourceHandle(search), opts);
  }

  async searchIndexWriter(kv, opts = null) {
    return _BINDINGS.searchIndexWriter(resourceHandle(kv), opts);
  }

  async searchWrite(writer, docRef, docText) {
    return _BINDINGS.searchWrite(resourceHandle(writer), docRef, docText);
  }

  async searchCommit(writer) {
    return toJsResult(await _BINDINGS.searchCommit(resourceHandle(writer)));
  }

  async newVectorIndex(kv, opts) {
    return _BINDINGS.newVectorIndex(resourceHandle(kv), opts);
  }

  async closeVectorIndex(index) {
    await _BINDINGS.closeVectorIndex(resourceHandle(index));
  }

  async vectorIndexClosed(index) {
    return _BINDINGS.vectorIndexClosed(resourceHandle(index));
  }

  async vectorAddVec(index, vecRef, vecData) {
    return toJsResult(await _BINDINGS.vectorAddVec(resourceHandle(index), vecRef, vecData));
  }

  async vectorRemoveVec(index, vecRef) {
    return toJsResult(await _BINDINGS.vectorRemoveVec(resourceHandle(index), vecRef));
  }

  async vectorIndexed(index, vecRef) {
    return _BINDINGS.vectorIndexed(resourceHandle(index), vecRef);
  }

  async vectorSearch(index, queryVec, opts = null) {
    return toJsResult(await _BINDINGS.vectorSearch(resourceHandle(index), queryVec, opts));
  }

  async vectorReIndex(index, opts = null) {
    return _BINDINGS.vectorReIndex(resourceHandle(index), opts);
  }

  async vectorClear(index) {
    return toJsResult(await _BINDINGS.vectorClear(resourceHandle(index)));
  }

  async vectorForceCheckpoint(index) {
    return toJsResult(await _BINDINGS.vectorForceCheckpoint(resourceHandle(index)));
  }

  async vectorInfo(index) {
    return toJsResult(await _BINDINGS.vectorInfo(resourceHandle(index)));
  }

  async vectorCheckpointState(index) {
    return toJsResult(await _BINDINGS.vectorCheckpointState(resourceHandle(index)));
  }

  async newClient(uri, opts = null) {
    return _BINDINGS.newClient(uri, opts);
  }

  async closeClient(handle) {
    await _BINDINGS.closeClient(resourceHandle(handle));
  }

  async clientDisconnected(handle) {
    return _BINDINGS.clientDisconnected(resourceHandle(handle));
  }

  async readEdn(edn) {
    return _BINDINGS.readEdn(edn);
  }

  async writeEdn(value) {
    return _BINDINGS.writeEdn(value);
  }

  async keyword(value) {
    return _BINDINGS.keyword(value);
  }

  async symbol(value) {
    return _BINDINGS.symbol(value);
  }

  async schema(schema) {
    return _BINDINGS.schema(schema);
  }

  async options(opts) {
    return _BINDINGS.options(opts);
  }

  async udfDescriptor(descriptor) {
    return _BINDINGS.udfDescriptor(descriptor);
  }

  async createUdfRegistry() {
    return _BINDINGS.createUdfRegistry();
  }

  async registerUdf(registry, descriptor, fn) {
    return _BINDINGS.registerUdf(registry, descriptor, fn);
  }

  async unregisterUdf(registry, descriptor) {
    return _BINDINGS.unregisterUdf(registry, descriptor);
  }

  async registeredUdf(registry, descriptor) {
    return _BINDINGS.registeredUdf(registry, descriptor);
  }

  async renameMap(renameMap) {
    return _BINDINGS.renameMap(renameMap);
  }

  async deleteAttrs(attrs) {
    return _BINDINGS.deleteAttrs(attrs);
  }

  async lookupRef(value) {
    return _BINDINGS.lookupRef(value);
  }

  async datom(e, attr, value, tx = undefined, added = undefined) {
    return datomTuple(e, attr, value, tx, added);
  }

  async txData(txData) {
    return _BINDINGS.txData(txData);
  }

  async kvTxs(txs, kType = null, vType = null) {
    if (kType === null && vType === null) {
      return _BINDINGS.kvTxs(txs);
    }
    return _BINDINGS.kvTxsWithTypes(txs, kType, vType);
  }

  async kvInput(value, type) {
    return _BINDINGS.kvInput(value, type);
  }

  async kvRange(range, type) {
    return _BINDINGS.kvRange(range, type);
  }

  async kvType(value) {
    return _BINDINGS.kvType(value);
  }

  async databaseType(value) {
    return _BINDINGS.databaseType(value);
  }

  async role(role) {
    return _BINDINGS.role(role);
  }

  async permissionKeyword(value) {
    return _BINDINGS.permissionKeyword(value);
  }

  async permissionTarget(objectType, target) {
    return _BINDINGS.permissionTarget(objectType, target);
  }
}

const RAW_INTEROP = new RawInterop();

export function interop() {
  return RAW_INTEROP;
}
